LocalApi.py

```diff
@@ -0,0 +1,250 @@
+import random
+import logging
+import bz2
+import pickle
+import itertools
+import xml.etree.ElementTree as xml
+import mwparserfromhell
+from titlecase import titlecase
+from os.path import getsize
+from IWikiApi import IWikiApi
+
+log = logging.getLogger(__name__)
+
+class WikipediaIndexFile:
+    """
+    A class to represent the index file provided with the data dump
+    The index file is a sequence of lines, each of the form:
+    BZ2_OFFSET:ARTICLE_ID:ARTICLE_TITLE
+
+    When we read the index file, we want to keep track of the ARTICLE_TITLE and the BZ2_OFFSET.
+    Articles are arranged in groups with the same BZ2_OFFSET,
+    and we need to know the BZ2_OFFSET of the group and the BZ2_OFFSET of the group after.
+
+    This allows us to take a clipping of the BZ2 file to decompress.
+    """
+    def __init__(self, index_file, xml_file_size):
+        self.index_file = index_file
+        self.xml_file_size = xml_file_size
+        self.articles = {}
+        self.article_titles = []
+
+    def manual_load(self):
+        """
+        Read through the index file and build up a map from article titles to tuples of the form (chunk_start, chunk_end).
+        (chunk_start, chunk_end) are the bytes that you would need to extract from the xml.bz2 file to find the chunk
+        containing a given page. Note that chunks may contain more than one article.
+
+        Store the current chunk region and titles in a buffer, when we find a new chunk we know the beginning 
+        and end of the previous chunk so we can add them to the dictionary.
+        """
+        buffer_region_start = 0
+        buffer_titles = []
+        with open(self.index_file, "r") as index_file:
+            for line in index_file:
+                [line_region_start, _id, *split_name] = line.split(":")
+                line_article_name = ":".join(split_name).strip()
+                if line_region_start != buffer_region_start:
+                    for name in buffer_titles:
+                        self.articles[name] = (int(buffer_region_start), int(line_region_start))
+                    buffer_region_start = line_region_start
+                    buffer_titles = []
+                buffer_titles.append(line_article_name)
+        for name in buffer_titles:
+            self.articles[name] = (int(buffer_region_start), int(self.xml_file_size))
+
+    def load(self):
+        """
+        Load the wikipedia index file into memory.
+        If an articles.pickle.tmp exists, we simply load that (~12 seconds)
+        If the pickle file doesn't exist, we manually load the index by parsing the text file (~43 seconds)
+        """
+        try:
+            with open("articles.pickle.tmp", "rb") as articles_pickle_file:
```

Comment:
You should probably check if the file exists explicitly here instead of using errors for control flow https://stackoverflow.com/questions/82831/how-do-i-check-whether-a-file-exists-without-exceptions

Reply:
I mean, I can do it with the explicit file check, but I also want to catch the case where the `pickle.load` call fails (which can happen if you manage to ctrl-c the program while it's writing the pickle file). Using the try-catch ensure that no matter what, if the loading of the pickle file fails, we just do it manually